tools/systemd-generator: Multi-interface and multi-host support (#5689)
zeek/zeek: doc/advanced/deployment/systemd.rst

.. _deploy-systemd:

============================
Deploying Zeek using systemd
============================

.. versionadded:: 8.1

Zeek includes a `systemd generator <https://www.freedesktop.org/software/systemd/man/latest/systemd.generator.html>`_
that reads the ``<PREFIX>/etc/zeek/zeek.conf`` configuration file and instantiates systemd

> unit files for the manager, logger, proxy and worker processes accordingly.

Reviewer: Maybe a tiny bit more context for people just passing through?
```suggestion
unit files for the manager, logger, proxy and worker processes accordingly.
systemd automatically runs generators during system startup, or on demand (see below).
```

The generator is installed at ``<PREFIX>/bin/zeek-systemd-generator``.

.. note::

   ``zeek.conf`` file's focus is an opinionated Zeek deployment tailored for
   appliance or container use cases. It presents an alternative to a

>    :ref:`ZeekControl <cluster-configuration>` managed deployment.

Reviewer: Could we not use "opinionated"? I think it has mostly bad connotations — "it's quirky", "it's partial", "I'm not willing to debate your takes", etc. That might just be me, but more importantly I think that's not actually true here! We did spend a bunch of time thinking through pros/cons etc.

An alternative would be to say that this approach prioritizes system-level cluster orchestration, not Zeek-side configuration — which ZeekControl continues to offer.

Author: > Could we not use "opinionated"? 

Yeah, certainly! Thanks for pointing this out, I'll ditch that usage and made a note to avoid it in the future.

Installation
============

After installing Zeek on your system, link Zeek's systemd generator into one
of the standard generator directories:

.. code:: console

>    # ln -s <PREFIX>/bin/zeek-systemd-generator /etc/systemd/system-generators/

Reviewer: Fyi, I hit some hiccups when I first tried this:
* I had to create the generator directory — none of the ones shown on the freedesktop.org page existed. That gave me some pause (but worked fine).
* I had a user-local install of the Zeek tree: the symlink was from `~/inst/opt/zeek/bin/zeek-systemd-generator`. SELinux balks at this, at two levels: (1) running the generator, (2) loading the config. The journal has the usual instructions to fix this, but it's subtle because without looking in the journal there's no output — you just don't get unit files. These hurdles go away with a system-level install of Zeek in `/usr/local`. (I have not tried `/opt`.)

So we should probably add something here to warn about this kind of stuff.

(This was on Fedora, btw.)

Author: Phew ;-) I made a note about looking at `journalctl` and also how to run the generator by hand as a pre-check as well as mentioning AppArmor and SELinux. I don't think we should document concrete strategies (yet?).

Reviewer: I agree, yeah.

Single Host Walkthrough
=======================

For a small Zeek deployment listening with 2 workers on eth1, using
AF_PACKET for flow-balancing, plus running the manager and one logger

> and one proxy process, the following minimal configuration is sufficient:

Reviewer: Just simplifying a little:
```suggestion
For a small Zeek deployment running two AF_PACKET worker processes on eth1, 
one logger, one proxy, and the manager, the following configuration is sufficient:
```

.. code:: ini

    interface = af_packet::eth1
    workers = 2

This minimal format is termed the section-less configuration style.
See the :ref:`systemd_multiple_interfaces` section for the more verbose INI-style
configuration that is required when you want to listen on multiple interfaces.

After running ``systemctl daemon-reload``, which executes ``zeek-systemd-generator``,
the generated unit files are found in the ``/run/systemd/generator`` directory of

> the system where they can be inspected:

Reviewer: ```suggestion
the system:
```
(less passive)

    Apr 24 13:42:15 tinkyx1 zeek-worker-2[2602816]: listening on eth1
    Apr 24 13:42:15 tinkyx1 zeek-worker-1[2602825]: listening on eth1
    Apr 24 13:46:44 tinkyx1 systemd[1]: zeek-worker@1.service: Main process exited, code=dumped, status=6/ABRT
    Apr 24 13:46:44 tinkyx1 systemd[1]: zeek-worker@1.service: Failed with result 'core-dump'.
    Apr 24 13:46:45 tinkyx1 systemd[1]: zeek-worker@1.service: Scheduled restart job, restart counter is at 2.
    Apr 24 13:46:45 tinkyx1 systemd[1]: Starting zeek-worker@1.service - Zeek Worker 1...
    Apr 24 13:46:45 tinkyx1 systemd[1]: Started zeek-worker@1.service - Zeek Worker 1.
    Apr 24 13:46:45 tinkyx1 zeek-worker-1[2603512]: listening on eth1

This output shows zeek-worker@1 terminating due to a SIGABRT (here, ``kill -SIGABRT``
was used for demonstration purposes) and systemd automatically restarting the process
and reporting the current restart counter value. Use ``systemctl show zeek-worker@1``
to inspect this counter and various other details.

For debugging of crashes and coredump handling in general, we recommend installing
`systemd-coredump <https://www.freedesktop.org/software/systemd/man/latest/systemd-coredump.html>`_

> and setting up ``/etc/systemd/coredump.conf`` as needed.

Reviewer: ```suggestion
if your distro doesn't automatically do so, and setting up ``/etc/systemd/coredump.conf`` as needed.
```
At least on Fedora `coredumpctl` ships automatically.

     Control Group: /zeek.slice/zeek-workers.slice/zeek-worker@1.service
              Unit: zeek-worker@1.service
             Slice: zeek-workers.slice
           Boot ID: 0d20f470e7a9413d92f7324c27866992
        Machine ID: c9dbccfd439e43b598fa6153f5fa9e3e
          Hostname: tinkyx1
           Storage: /var/lib/systemd/coredump/core.zeek.997.0d20f470e7a9413d92f7324c27866992.2603455.1777031203000000.zst (present)
      Size on Disk: 10.4M
           Message: Process 2603455 (zeek) of user 997 dumped core.
    
                    Module libzstd.so.1 from deb libzstd-1.5.5+dfsg2-2build1.1.amd64
                    Module libsystemd.so.0 from deb systemd-255.4-1ubuntu8.15.amd64
                    Module libgcc_s.so.1 from deb gcc-14-14.2.0-4ubuntu2~24.04.1.amd64

.. _systemd_multiple_interfaces:

Multiple Interfaces
===================

.. versionadded:: 9.0

To monitor multiple interfaces, the ``zeek.conf`` file supports INI-style sections
where there is one section per interface. Each interface section name includes
a **tag** that's used in systemd's unit files, working directories, and also the
cluster node name. The tags used in the sample configuration below are ``eth1``

> and ``eth2``.

Reviewer: Do you think the notion of a tag matters to the user? If we just just assume that the user will usually want this to be the interface name, that could then become something like
```suggestion
To monitor multiple interfaces, add an INI-style section ``[interface <name>]``
to ``zeek.conf`` for each interface. The generator uses the interface's name
in systemd's unit files, working directories, and also the cluster node name.
```
When would it _not_ be an interface?

Author: > When would it not be an interface?

I was balancing between the resulting interface names that individual workers use (*after* interpolation of the `interface =` string) and the tag/name of the section. I do like `name` in the section as you propose a lot more than `tag`! It's true that the interpolated string isn't really exposed so this works. I'll switch/change this.

Author: Oh, and just to give a few examples for why I think the tag/name abstraction is useful:

> When would it not be an interface?

```
[interface internal-traffic]
interface = af_packet::eth1

[interface external-traffic]
interface = af_packet::eth2

[interface netmap-eth0]
interface = netmap::eth0{${worker_index0}

[interface vxlan]
interface = udp::0.0.0.0:4789:vxlan

[interface geneve]
interface = udp::[::]:6081:geneve

[interface napatech]
interface = napatech::${worker_index0}

# Four workers in a custom_lb group listening on separate
# interfaces: `veth1`, `veth2`, `veth3` and `veth4`.
# This is useful if you build a custom load-balancer/packet-broker.
# For example, a bond interface with balance-xor
# and xmit_hash_policy=encap3+4 and four slaves
# attached would do, but could also be an XDP program
# that flow-balances by forwarding packets to created
# interfaces.
[interface xdp-lb]
interface af_packet::veth${worker_index}
workers = 4
```

Not sure all of the above make sense, but the separation between actual `-i` value and a descriptive name seems very nice and avoids coming up with escaping/normalization patterns (the interface section name is used in working directories and cluster-node names).

Author: I've made this now be "name" in code and docs. Tag or identifier could  help to separate from the actual `-i` argument that'll be used, but "name" reads nicer.

Reviewer: Those examples are great. It makes a lot more sense to me now.

Author: I've added some of these examples to the default config as comments.

When using the INI-style format, non-interface option must be placed into

> the ``[zeek]`` section, conventionally put at the top.

Reviewer: ```suggestion
When using the INI-style format, place non-interface options into
the ``[zeek]`` section, conventionally at the top of the file:
```

.. code:: ini

    [zeek]
    loggers = 3
    proxies = 7

    [interface eth1]

>     interface = af_packet::eth1

Reviewer: I wondered about defaults at this point. This pattern ...
```
    [interface eth1]
    interface = af_packet::eth1
```
will be very common, right? Should that value (`af_packet::<name>`) be the default?

Actually, are there any defaults? (I later got to the code and don't think so, but I'm not through yet.)

Author: > will be very common, right? Should that value (af_packet::<name>) be the default?

:-)

I had pondered, too. But pushing back here. `af_packet` is Linux specific and `zeek.conf` is not, so I'd avoid it as it'd mean a platform specific default. The `interface` entry is also what's directly passed to `-i` after variable interpolation.

Seems also useful to expose users (even new ones) to packet source plugins. It's a fairly important part that makes a live Zeek system work correctly.

A complex `zeek.conf` should always be generated using some configuration management tool.

Reviewer: Okay. We can also always add this later. I chuckled when I saw this because it directly reminded me of
```
[manager]
role = manager
```
in the management framework, and feedback then (Dop, Smoot) was that it'll usually be redundant. (And I never fixed it. :-)

Author: The section-less format would not have:
```
manager = 1
loggers = 1
proxies = 1
workers = 4
interface = af_packet::eth1
```

And for INI-style, the section name is used in working directories, unit names and cluster node names rather than actually influencing the effective interface used by a worker, so seems a separate enough use to keep it explicit.

```
[zeek]
manager = 1
loggers = 1
proxies = 1

[interface eth1]
interface = af_packet::eth1
workers = 4
```

Seems also okay to type by hand in demos and trivial to have a jinja template or some such elsewhere.

    workers = 4
    workers_cpu_list = 4-7
    worker_args = AF_Packet::fanout_id=42

    [interface eth2]
    interface = af_packet::eth2
    workers = 4
    workers_cpu_list = 8-11
    worker_args = AF_Packet::fanout_id=4711
    worker_env =
      LD_PRELOAD=/usr/local/lib/libjemalloc.so
      MALLOC_CONF=prof:true,prof_prefix:jeprof.out,prof_final:true,lg_prof_interval:26

With AF_PACKET, the respective workers have to use different fanout groups, so we need
to pass an explicit ``AF_Packet::fanout_id`` setting via the ``worker_args`` option.
This configuration also pins workers sequentially onto CPUs 4 through 11
and enables jemalloc profiling for all workers listening on eth2 via the
``worker_env`` option.

> You will find that the worker unit names include the interface tags when

Reviewer: ```suggestion
You will find that the worker unit names include the interface names when
```
(if we end up not using the notion of tags)

configured like this. I.e., instead of ``zeek-worker@1``, the first worker
for each interface is ``zeek-worker-eth1@1`` and ``zeek-worker-eth2@2``.
The ``CLUSTER_NODE`` and :zeek:see:`Cluster::node` values change accordingly:
``worker-1`` is ``worker-eth1-1`` and ``worker-eth2-1``. You can see these
values in the generated ``cluster-layout.zeek`` file.

Interface Templating
====================

> The ``interface`` option supports templating. For example, selecting a

Reviewer: This sounds a bit like there are variables other than `worker_index0`. Are there?

Reviewer: To answer my own question, I found the `vars` array logic  in `systemd_write_units()`. So it's probably worth spelling them out?

Author: The whole config needs some sort of reference documentation. I'd like to postpone this though and land this first. I'd guess there might be a few more tweaks and the PR here is already pretty big.

But added a table about the available variables for `interface =` specifically now.

Reviewer: Cool, thank you!

worker-specific Napatech stream for the interface looks as follows:

.. code:: ini

    [interface napatech]
    interface = napatech::${worker_index0}
    workers = 32

The ``${worker_index0}`` variable expands to the zero-based index of each worker
for this interface (0 through 31) and so every worker receives an individual
interface ``napatech::0``, ``napatech::1`` through ``napatech::31``. This can
also be useful to select a specific NETMAP pipe per worker.

.. code:: ini

    [interface eth1]
    interface = netmap::eth1{${worker_index0}
    workers = 32

Multi Host Walkthrough
======================

.. versionadded:: 9.0

.. note::

   Multi-host support using the zeek.conf is still being refined. We plan
   to finalize this towards Zeek 10.0. We welcome testing and feedback!

You can think of a host as an actual physical or virtual system, but it could
also be just a Linux container with its own network namespace. Any environment
with a dedicated networking stack and hostname can be considered a host in the

> following paragraphs.

Reviewer: This section gets pretty complex, so I'd suggest shortening it. I'd mention two things up-front: (1) we assume that you have a compatible Zeek installation on each host, (2) each host has functional name resolution for each host in the cluster (basically what you're saying in line 297-302).

Author: Yep! This was a long detour. I moved the cluster-layout chunks into a "Cluster Layout Notes" section as some pieces seemed useful, but it's a bit rough.

To create a Zeek cluster that spans multiple hosts, the format of the ``zeek.conf``
file does not change. However, you place the configuration files of all hosts of the
cluster at ``<PREFIX>/etc/zeek/cluster/<hostname>.zeek.conf`` on each host.
That means that every host has visibility into every other host's config. Use some
configuration management tool to keep the ``<PREFIX>/etc/zeek/cluster`` directories
in sync.

When no ``<PREFIX>/etc/zeek/zeek.conf`` file for a single-host configuration exists,
the ``zeek-systemd-generator`` will attempt to open ``<PREFIX>/etc/zeek/cluster/<hostname>.zeek.conf``
using the hostname of the system it is running on. If the file is found, it'll use

> it to configure the appropriate systemd services using that specific config.

Reviewer: At this point I'd stop and just shorten to:
```suggestion
it to configure the appropriate systemd services using that specific config
and generate ``cluster-layout.zeek``.
```
If you'd rather keep more detail, I'd move it to a dedicated section about `cluster-layout.zeek`.

When executing the steps in the generated ``zeek-setup.service``, a ``cluster-layout.zeek``
containing all hosts and nodes is generated based on all of the ``*.zeek.conf`` files
within the ``<PREFIX>/etc/zeek/cluster`` directory.

This in an excerpt of the generated ``zeek-setup.service``, specifically the invocation
of the ``zeek-cluster-layout-generator`` tool.

.. code:: console

   # cat /run/systemd/generator/zeek-setup.service
   # Auto-generated, do not edit. Use drop-in files instead!
   ...
   [Service]
   Type=oneshot
   User=
   Group=
   WorkingDirectory=
   ExecStart=mkdir -p /usr/local/zeek/var/spool/zeek/generated-scripts
   ExecStart=/usr/local/zeek/bin/zeek-cluster-layout-generator -C /usr/local/zeek/etc/zeek/cluster -o /usr/local/zeek/var/spool/zeek/generated-scripts/cluster-layout.zeek

As of version 9.0, each host needs to be statically aware of all other hosts in
a cluster and able to resolve their IP addresses though the hostname as used in
the configuration filenames. Adding the required entries into ``/etc/hosts`` on
the respective systems is one way to achieve this, DNS would be another. Container

> orchestrators usually allow to specify hostnames for containers or pods and

Reviewer: ```suggestion
orchestrators usually allow specifying hostnames for containers or pods, and
```

provide name resolution services as well.

You can point the ``zeek-cluster-layout-generator`` executable at a
``<PREFIX>/etc/zeek/cluster`` directory and inspect the generated layout.
If you do not specify a ``cluster_address`` option within the individual
``<hostname>.zeek.conf`` files, the generated ``cluster-layout.zeek`` file
contains extra script code to use Zeek's DNS functionality to lookup the
addresses of the involved hosts.

.. code:: console

    # zeek-cluster-layout-generator -C ./cluster/
    # Auto-generated by zeek-cluster-layout-generator
    ...

    const hosts: table[string] of Host &ordered &redef;

    redef Cluster::hosts += {

>             ["c-mgr"] = [$ip=(blocking_lookup_hostname("c-mgr") as vector of addr)[0]],

Reviewer: Btw ... this reminded me that we should move the `Node` record away from address-only `$ip` to something string-based, perhaps `$host`. 9.1+, I guess.

Author: :-)

I'm not entirely sure here. I realize that with ZeroMQ or NATS, neither `$ip` nor `$host` would be all that interesting for pubsub functionality anymore.  One can connect to `tcp://mgr:5555` and `tcp://mgr:5556` and interact with the cluster (and similarly with a `nats_url`).

That's all to say that instead of moving `$ip` to `$host`, may also be able to remove these altogether.

    };

    redef Cluster::nodes += {
        ["manager"] = [$node_type=Cluster::MANAGER, $ip=hosts["c-mgr"]$ip, $p=27760/tcp, $metrics_port=9991/tcp],
        ["c-mgr-logger-1"] = [$node_type=Cluster::LOGGER, $ip=hosts["c-mgr"]$ip, $p=27761/tcp, $manager="manager", $metrics_port=9992/tcp],
    }

This means that each Zeek process will determine the addresses of all other
hosts in a cluster when loading the generated ``cluster-layout.zeek`` script
during startup. That's why it is important to have a functional ``/etc/hosts``
or DNS setup. Alternatively, set ``cluster_address`` to the host's address
in the respective ``<hostname>.zeek.conf``.

You can debug the resulting layout using Zeek itself:

.. code:: console

    # zeek -b <PREFIX>/var/spool/zeek/generated-scripts/cluster-layout.zeek -e 'print Cluster::nodes'
    {
    [c-w-02-worker-vxlan-4] = [node_type=Cluster::WORKER, ip=fd00:dead:beef::12, zone_id=, p=0/unknown, manager=manager, id=<uninitialized>, metrics_port=9994/tcp],
    [c-w-01-worker-vxlan-3] = [node_type=Cluster::WORKER, ip=fd00:dead:beef::11, zone_id=, p=0/unknown, manager=manager, id=<uninitialized>, metrics_port=9993/tcp],
    ...
    }

> For a minimal three host cluster, where host ``c-mgr`` runs manager, 3 loggers,

Reviewer: And I'd resume here (from line 275), with this example.

2 proxies and the archiver and two hosts ``c-w-01`` and ``c-w-02`` that each
run 4 workers, listening on their eth1 interface using AF_PACKET, the following
configuration files are sufficient.

.. code:: ini

   # <PREFIX>/etc/zeek/cluster/c-mgr.zeek.conf
   [zeek]
   manager  = 1
   loggers  = 3
   proxies  = 2
   archiver = 1
   
   cluster_backend_args = misc/zeromq-multi-host-auto-setup

.. code:: ini

   # <PREFIX>/etc/zeek/cluster/c-w-01.zeek.conf
   [zeek]
   loggers  = 0
   proxies  = 0
   manager  = 0

>    archiver = 0

Reviewer: This again had me wondering about defaults — would be nice to not need to say this?

Author: Yeah. I've been forth and back. They actually default to `1` to make everything work out of the box on a single-host.

I had pondered changing the defaults to `0` when the config has a hostname (for multi-host), or an extra option to set them all to 0, but then it seemed this should almost always be configuration management maintained in a production environment, so unclear what this is optimizing for (and the defaults are optimized for single-host demo "just" working)

```
[zeek]
only_workers = true

[zeek]
workers = only

[zeek]
only_workers
```

Reviewer: Mhmm. I would make the single-host scenario more explicit instead. I'm glad this came up again here because I found that two-liner of 
```
    interface = af_packet::eth1
    workers = 2
```
in the single-host setting a bit extreme when I first saw it. Does that default to one logger? To manager-is-logger? To one proxy or a handful? Etc. Most people will also start from the default zeek.conf, which already lists the other node types anyway, just commented out. So if a minimal full cluster becomes
```
    interface = af_packet::eth1
    workers = 2
    logger = 1
    proxy = 1
    manager = 1
``` 
or perhaps
```
    interface = af_packet::eth1
    workers = 2
    proxy = 1
    manager = 1
```
for manager-is-logger, I think that's pretty clean. `manager = 1` looks a bit pointless, but there's some future-proofing in this: we've already considered loggerless clusters, people have asked for standby managers, and perhaps you actually want to run experimental/short-lived setups that aren't complete clusters. Long story short, I think being explicit is good.

> so unclear what this is optimizing for

I mean, clutter? If you have a 10 machine cluster, you need to mention all the stuff they're each _not_ running roughly 10 times.

Author: > I'm glad this came up again here because I found that two-liner of [...]
> in the single-host setting a bit extreme when I first saw it. Does that default to one logger? To manager-is-logger? 

> I mean, clutter?  If you have a 10 machine cluster, you need to mention all the stuff they're each not running roughly 10 times.

Yeah, this is good. Thanks for the pushback! 

I've changed the defaults for the manager, proxies, loggers and workers options to be `0` and updated the tests. A bit of churn, but agree the explicitness feels much better. Only putting interface into zeek.conf doesn't work anymore, but that's alright. The `archiver` option defaults to `1 if (manager == 1 ||  loggers >= 1) else 0` if unset now.
